matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/network/Request.kt

```diff
@@ -19,38 +19,45 @@ package org.matrix.android.sdk.internal.network
-
-internal class Request<DATA : Any>(private val globalErrorReceiver: GlobalErrorReceiver?) {
-
-    var isRetryable = false
-    var initialDelay: Long = 100L
-    var maxDelay: Long = 10_000L
-    var maxRetryCount = Int.MAX_VALUE
-    private var currentRetryCount = 0
-    private var currentDelay = initialDelay
-    lateinit var apiCall: Call<DATA>
+/**
+ * Execute a request from the requestBlock and handle some of the Exception it could generate
+ *
+ * @param globalErrorReceiver will be use to notify error such as invalid token error. See [GlobalError]
+ * @param canRetry if set to true, the request will be executed again in case of error, after a delay
+ * @param initialDelayBeforeRetry the first delay to wait before a request is retried. Will be doubled after each retry
+ * @param maxDelayBeforeRetry the max delay to wait before a retry
+ * @param maxRetriesCount the max number of retries
+ * @param requestBlock a suspend lambda to perform the network request
+ */
+internal suspend inline fun <DATA> executeRequest(globalErrorReceiver: GlobalErrorReceiver?,
+                                                  canRetry: Boolean = false,
+                                                  initialDelayBeforeRetry: Long = 100L,
+                                                  maxDelayBeforeRetry: Long = 10_000L,
+                                                  maxRetriesCount: Int = Int.MAX_VALUE,
+                                                  noinline requestBlock: suspend () -> DATA): DATA {
+    var currentRetryCount = 0
+    var currentDelay = initialDelayBeforeRetry
 
-    suspend fun execute(): DATA {
-        return try {
-            val response = apiCall.clone().awaitResponse()
-            if (response.isSuccessful) {
-                response.body()
-                        ?: throw IllegalStateException("The request returned a null body")
-            } else {
-                throw response.toFailure(globalErrorReceiver)
+    while (true) {
+        try {
+            return requestBlock()
+        } catch (throwable: Throwable) {
+            val exception = when (throwable) {
+                is KotlinNullPointerException -> IllegalStateException("The request returned a null body")
+                is HttpException              -> throwable.toFailure(globalErrorReceiver)
+                else                          -> throwable
             }
-        } catch (exception: Throwable) {
-            // Log some details about the request which has failed
-            Timber.e("Exception when executing request ${apiCall.request().method} ${apiCall.request().url.toString().substringBefore("?")}")
+
+            // Log some details about the request which has failed. This is less useful than before...
+            // Timber.e("Exception when executing request ${apiCall.request().method} ${apiCall.request().url.toString().substringBefore("?")}")
+            Timber.e("Exception when executing request")
```

Comment:
When it's an HttpException we can grab the response and request I think?

Reply:
Sadly, no